text/3830-dedented-string-literals.md

```diff
@@ -0,0 +1,1569 @@
+        create table student(
+            id int primary key,
+            name text
+        )
+    ");
+}
+```
+
+Which outputs (using `^` to mark the beginning of a line, and `·` to mark a leading space):
+
+```sql
+^
+^········create table student(
+^············id int primary key,
+^············name text
+^········)
+^····
+^
+```
+
+The output is formatted in an unconventional way, containing excessive leading whitespace.
+
+The alternative allows for a sane output, but at the cost of making the code less readable:
+
+### Sacrifice readability of the source code
+
+In order for the output to be more sensible, we must sacrifice readability of the source code:
+
+```rust
+fn main() {
+    println!(
+        "\
+create table student(
+    id int primary key,
+    name text
+)");
+}
+```
+
+The above example would output the expected:
+
+```sql
+create table student(
+    id int primary key,
+    name text
+)
+```
+
+But the improvement in output comes at a cost:
+
+1. We now have to escape the first newline:
+
+   ```diff
+   fn main() {
+       println!(
+   +       "\
+   create table student(
+       id int primary key,
+       name text
+   )");
+   }
+   ```
+
+   This is not possible to do in raw strings, so the output ends up looking even worse for them, with indentation of the outer SQL statement being larger in the source code than the inner statement:
+
+   ```diff
+   fn main() {
+       println!(
+   +       r#"create table student(
+       id int primary key,
+       name text
+   )"#);
+   ```
+
+2. The SQL statement does not have any indentation in reference to the surrounding code.
+
+   This is contrary to how we would normally write code, with indentation 1 level deeper than the surrounding.
+
+   ```diff
+   fn main() {
+       println!(
+           "\
+   +create table student(
+   +   id int primary key,
+   +   name text
+   +)");
+   }
+   ```
+
+   This makes it confusing to tell which scope the string belongs to. This is especially true when there are multile scopes involved:
+
+   ```rs
+   fn main() {
+       {
+           println!(
+               "\
+   create table student(
+      id int primary key,
+      name text
+   )");
+       }
+       println!(
+           "\
+   create table student(
+      id int primary key,
+      name text
+   )");
+       {
+           {
+               println!(
+                   "\
+   create table student(
+      id int primary key,
+      name text
+   )");
+            
+           } 
+       }
+   }
+   ```
+
+   All of the strings end up on the same level, despite them being in different scopes.
+
+3. The closing double-quote must be put at the beginning of the line, in order not to introduce trailing whitespace:
+
+   ```diff
+   fn main() {
+       println!(
+           "\
+   create table student(
+       id int primary key,
+       name text
+   +)");
+   }
+   ```
+
+As you can see, we have to choose one or the other. In either case we have to give something up.
```

Comment:
Another option worth mentioning is:

```rs
fn main() {
    println!(concat!(
        "create table student(\n",
        "    id int primary key,\n",
        "    name text,\n",
        ")\n",
    ));
}
```

This formats and indents nicely, and produces correct output, but has the disadvantage of looking very noisy. Additionally, `{variable_name}` formatting parameters are not supported, since the format string is expanded by a macro.